### ISSUE TITLE
Recreate the Device Manager gRPC server if failed

### ISSUE THREAD

**Issue by SergeyKanzhelev** [kubernetes/kubernetes#128167]
We need to ensure Device Plugin infrastracture is reliable by implementing the following:

1. Retry to start the server if it failed (see https://github.com/kubernetes/kubernetes/pull/125513 for places where the gRPC server may fail).
2. If server is not up, integrate it as a source for the kubelet health status (see https://github.com/kubernetes/kubernetes/issues/127460)

Retry should be reasonable easy to implement. Instead of a goroutine that simply started the Server, an advanced goroutine that Serve and retries can be used. I think it would be best to also try to re-create the socket if it was deleted. The retry logic must be unit tested as well as e2e_node test should be added.

Integrating with the health checking will require implementing an interface that will report "unhealthy". The important part here is that the health checking must not involve any lock or calculations. A simple bool value check should be a preferred way to do the health check.

/cc @ffromani @bart0sh 

/help
/kind feature
/sig node
/priority backlog
/area kubelet

**Comment by ffromani** [comment]
I agree. There's also the general trend about retry device-related pod admission failures: https://github.com/kubernetes/kubernetes/issues/128043#issuecomment-2420770273

**Comment by zhifei92** [comment]
@SergeyKanzhelev  Hi, I would like to work on this issue. I have a few points to confirm:

1. Does `I think it would be best to also try to re-create the socket if it was deleted.` mean that we need to check whether the socket needs to be recreated before each retry of `s.grpc.Serve(ln)`? If so, could you explain under what circumstances the socket might be deleted?
2. Do you suggest that the two points you mentioned—retry and health check—should be separated into two different PRs?

Thanks.

**Comment by SergeyKanzhelev** [comment]
> * Does `I think it would be best to also try to re-create the socket if it was deleted.` mean that we need to check whether the socket needs to be recreated before each retry of `s.grpc.Serve(ln)`? If so, could you explain under what circumstances the socket might be deleted?

I am imagining somebody cleaned up the whole directory by mistake. Also potentially somebody started a second kubelet by mistake soemehow. Both are not super probable. But I think it may be the best option.

> * Do you suggest that the two points you mentioned—retry and health check—should be separated into two different PRs?

Yes, keep those two separate. You can even do it in three PRs - one with re-creating the connection can be implemented separately.

**Comment by zhifei92** [comment]
@SergeyKanzhelev  Thank you for your response. Let's confirm the three requirements again:

1. Retry when `grpc.Serve` fails, and implement unit tests and e2e_node tests.
2. Integrate with health checks.
3. Monitor the socket; if it is deleted, recreate the Device Manager gRPC server.

/assign

**Comment by zhifei92** [comment]
@tunedev is working on the first part.
> 1. Retry when `grpc.Serve` fails, and implement unit tests and e2e_node tests.

@zhifei92 will do the second.
> 2. Integrate with health checks.

@tunedev For the third part, please leave a comment if you wish to continue.

**Comment by SergeyKanzhelev** [comment]
@zhifei92 feel free to take over other PRs. I don't see them moving.

**Comment by zhifei92** [comment]
> @zhifei92 feel free to take over other PRs. I don't see them moving.

okay, thank you.

**Comment by tunedev** [comment]
Hi @SergeyKanzhelev, @zhifei92

I'm following up on this issue and have reviewed the implementation in PR #128388 by @zhifei92. It provides a solid foundation for solving this problem IMO.

I have an existing PR #128366 for this issue, but I believe @zhifei92's approach is better. I want to help get this work merged.

@zhifei92: It looks like the PR has been inactive for a little while. Are you still planning on finishing it? I would be happy to collaborate with you by addressing the latest comments to get it ready for review.

Thanks!
